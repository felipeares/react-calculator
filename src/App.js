import React from "react";
import styles from "./App.module.css";

function App() {
  return <div className={styles.App}>hello</div>;
}

export default App;
