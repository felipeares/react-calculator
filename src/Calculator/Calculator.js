import React, { Component } from "react";
import classes from "./Calculator.module.css";
import Display from "./Display/Display";
import KeyPad from "./KeyPad/KeyPad";

class Calculator extends Component {
  state = {
    displayValue: "",
    firstNumber: "",
    firstOperator: "",
    secondNumber: ""
  };

  /* action handlers */
  numberPressedHandler = numberPressed => {
    // check which number to change
    // case no first operator
    if (this.state.firstOperator === "") {
      // change fist number
      const newFirstNumber = this.state.firstNumber + numberPressed.toString();
      this.setState({
        displayValue: newFirstNumber,
        firstNumber: newFirstNumber
      });
    }

    // case with first operaor
    else {
      // change second number
      const newSecondNumber =
        this.state.secondNumber + numberPressed.toString();
      this.setState({
        displayValue: newSecondNumber,
        secondNumber: newSecondNumber
      });
    }
  };

  operatorPressedHandler = operatorPressed => {
    // check state of calculator
    // case if no first number set yet. then do nothing
    if (this.state.firstNumber === "") return;

    // case if no second number set
    if (this.state.secondNumber === "") {
      // update de inner operator (first operator)
      this.setState({
        firstOperator: operatorPressed
      });
    }

    // case second number set, then it is time to update all
    if (this.state.secondNumber !== "") {
      // get the numbers parsed
      const firstNumberFloat = parseFloat(this.state.firstNumber);
      const secondNumberFloat = parseFloat(this.state.secondNumber);

      // build a mapped function
      const firstOperatorFunction = (a, b, operator) => {
        switch (operator) {
          case "+":
            return a + b;
          case "-":
            return a - b;
          case "*":
            return a * b;
          case "/":
            return a / b;
          case "=":
            return b;
          default:
            return b;
        }
      };

      // calculate new numbers and operators
      const newFirstNumber = firstOperatorFunction(
        firstNumberFloat,
        secondNumberFloat,
        this.state.firstOperator
      ).toString();

      const newFirstOperator = operatorPressed;
      const newSecondNumber = "";

      // set
      this.setState({
        displayValue: newFirstNumber,
        firstNumber: newFirstNumber,
        firstOperator: newFirstOperator,
        secondNumber: newSecondNumber
      });
    }
  };

  resetPressedHandler = () => {
    // back to initial state
    this.setState({
      displayValue: "",
      firstNumber: "",
      firstOperator: "",
      secondNumber: ""
    });
  };

  render() {
    return (
      <div className={classes.Calculator}>
        <Display displayValue={this.state.displayValue} />
        <KeyPad
          numberPressed={this.numberPressedHandler}
          operatorPressed={this.operatorPressedHandler}
          resetPressed={this.resetPressedHandler}
        />
      </div>
    );
  }
}

export default Calculator;
