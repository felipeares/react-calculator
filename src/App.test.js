import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import App from "./App";
import Calculator from "./Calculator/Calculator";

describe("App", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it("have not changed the UI", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render a div", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });

  it("should have a named App class", () => {
    expect(wrapper.find("div").hasClass("App")).toEqual(true);
  });

  it("should render the Calculator component", () => {
    expect(wrapper.containsMatchingElement(<Calculator />)).toEqual(true);
  });
});
