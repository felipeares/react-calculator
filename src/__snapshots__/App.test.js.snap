// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App have not changed the UI 1`] = `
<div
  className="App"
>
  <Calculator />
</div>
`;
